### DISCUSSION TITLE
how to get watermark with replication (ZK)

### DISCUSSION THREAD

**Discussion by petar-dambovaliev** [ClickHouse/ClickHouse#42495] [Q&A] [answered]
Hello,

How can i get a watermark from a replica, so i know if it has inserted rows since `some datetime`?
In the system tables, there is `inserts_oldest_time (DateTime) - See queue_oldest_time` but i am not sure what that means.
Did they mean to write `inserts_last_time`?!?

**Comment by mo-avatar** [comment]
May be you can check system.query_log to  see the latest  successful insert operation. To my knowledge, the inserts_oldest_time  thing mainly reflects the  log entry status in zookeeper and has little to do with real inserts queries.

**Reply by petar-dambovaliev** [reply]
This table does not specify which replica in each row. I can't use it.

**Reply by petar-dambovaliev** [reply]
Is there a way, when you run a query, on which  zookeeper server node it is running on?

**Reply by mo-avatar** [reply]
> This table does not specify which replica in each row. I can't use it.
actuallly it just reflects the info in the very replica of this clickhouse-server node, 

The replica concept is for clickhouse-server instance or more specifically replicatedmergetree instance.
zookeeper is just something like a mail box. it has nothing to do with insert operation itself but as a coordinator.

The rough process is like,  one replica of a certain replicated table get a insert and a part is generated, and it write an entry to zookeeper says "I got an insert operation and generated a part part_a ".
The other replica listen on this path, and if it find new record is generated, it will pull the part generated by the previous replica accordingly. so it really  doesn't matter which  zookeeper server node the entry is been written.   

But, still, you can find which zookeeper instance you are using by find which zookeeper instance is master when you perfrom inserts. The other zookeeper instance is mainly an backup of the master one.

**Reply by mo-avatar** [reply, marked as answer]
I just find that this the part_log table may help you with the problem, it reflects the part(data) changing process in a certain table, you can find the latest part change log in certain table(NEW_PART means insert,  DOWNLOAD_PART  means sync from peer replica), but still it  only says the infomation of the current clickhouse server_node.  but you can find the replica info in system.clusters.

**Reply by petar-dambovaliev** [reply]
`part_log` has unique data for each replica? Meaning, on every server, `part_log` will have different data, for that particular server.

**Reply by mo-avatar** [reply]
> `part_log` has unique data for each replica? Meaning, on every server, `part_log` will have different data, for that particular server.

Yes it is

**Reply by petar-dambovaliev** [reply]
This will do the job. Thank you very much.
